### ISSUE TITLE
Dynamically determine command for new pane?  Make default_prog a function?

### ISSUE THREAD

**Issue by lunaryorn** [wez/wezterm#1776]
**Describe the solution you'd like**
I'd like to use a function to determine the exact command to use for a new pane.  Currently `default_prog` is just a static table, so all new tabs and panes spawn the same program.  However I'd like to start a slightly different program for each new pane or tab.

For this purpose I'd like `default_prog` to be a function, as an alternative to a fixed table.  This function would receive a table with some information about the new tab/pane, and returns a table with the command to use.

Would you accept a pull request which also allows a function for `default_prog`?

See below for the why :slightly_smiling_face: 

**Describe alternatives you've considered**
Are there any other means to do this currently?

**Additional context**
I'd like to spawn every new shell into a dedicated systemd scope, to isolate tabs from each other and from the parent wezterm process. 

This puts each pane under separate memory/CPU accounting, and makes it become a separate target for the userspace OOM killer. If some program running in wezterm consumes an excessive amount of system memory systemd will now take down just the offending pane instead of the entire wezterm process (which otherwise forms the parent scope).

In and by itself that's easy by wrapping `systemd-run` around my shell in in `wezterm.lua`:

```lua
return {
  default_prog = {
    '/usr/bin/systemd-run',
    '--user',
    '--scope',
    '--description=Shell started by wezterm',
    '--same-dir',
    '--collect',
    '/usr/bin/fish'
  },
}
```

However there's a minor nuisance: The scope name gets randomly generated by `systemd-run` itself and isn't exactly descriptive:

```
$ systemctl --user status | rg -B 3 fish
           │ ├─run-r38b287d9251e426cb0d3e6997da4310c.scope
           │ │ ├─65490 /usr/bin/fish
           │ │ ├─65967 nvim .
           │ │ └─66044 /usr/bin/wl-copy --foreground --type text/plain
           │ ├─run-ra630afc556034014ab80e7a49745a99a.scope
           │ │ ├─66086 /usr/bin/fish
           │ │ ├─66378 systemctl --user status
           │ │ └─66379 rg -B 3 fish
```

I'd prefer if it was something like `wezterm-shell-$PID-$WINDOW-$TAB-$PANE-$RANDOM.scope` where `$PID`  is the PID of the current wezterm process, `$RANDOM` some random string for uniqueness, and the other variables unique identifiers for the Wezterm window, tab and pane running this shell.

I can pass a custom name for the new scope with the `--unit` option for `systemd-run`, but then systemd-run won't make that name unique.  In other words I have to generate a unique name for each new shell.  

But I can't, because `default_prog` is a static table.

**Comment by wez** [comment]
This is a neat idea!  There's some plumbing to facilitate this sort of thing; on Windows, we have a `WslDomain` that rewrites commands to spawn via `wsl.exe` that is conceptually similar to this request:

https://github.com/wez/wezterm/blob/4cd893335b3cb7b7986fc17547d0a9e68dc0320e/mux/src/domain.rs#L176-L212

When I wrote that, I had in mind the idea that we could extend it to also spawn into eg: docker containers, so it's not a huge stretch to consider also the container/isolation provided by systemd-run as part of this.

In terms of implementation, rather than making `default_prog` a function, I think we could add an event that is triggered by the `fixup_command` method and that is passed the command builder object.  That event could then return either `nil` (meaning no change) or an alternative command builder object.

That event would be synchronous, which means that some of the more powerful asynchronous helpers provided in the `wezterm` lua module would not be callable, but I think that would be sufficient for these purposes.

Note that in that context the pane_id is part of the environment set in the command builder.

**Comment by wez** [comment]
Here's an example of triggering a synchronous lua event; it emits the `mux-is-process-stateful` event and processes the result:

https://github.com/wez/wezterm/blob/4cd893335b3cb7b7986fc17547d0a9e68dc0320e/mux/src/localpane.rs#L383-L397

If you wanted to try making a PR for what I suggested above, I think the steps would be:

* A function to construct a https://wezfurlong.org/wezterm/config/lua/SpawnCommand.html from the command builder
* Dynamically construct the event name as `mux-local-domain-NAME-fixup-command` where `NAME` comes from `LocalDomain::name`.  That would allow defining multiple domains with different hooks in the future (eg: one domain per docker container, or something like that).  That would make the name something like `mux-local-domain-local-fixup-command` for the default `LocalDomain` of `"local"`.
* Emit the event using something like `(event_name, (spawn_command_we_built_earlier))` as parameters
* We expect the lua event handler to return either nil or a `SpawnCommand`
* If we get a `SpawnCommand`, make a new command builder using logic similar to:

https://github.com/wez/wezterm/blob/4cd893335b3cb7b7986fc17547d0a9e68dc0320e/wezterm-gui/src/termwindow/spawn.rs#L83-L102

Then in your wezterm config, you might have something like:

```lua
wezterm.on("mux-local-domain-local-fixup-command", function(cmd)
   local args = {"systemd-run", table.unpack(cmd.args)}
   return cmd
end)
```

**Comment by lunaryorn** [comment]
@wez Oh that escalated quickly :laughing:   But a new event is a nice idea; I see if I can make a PR about this.  Will take a while though, I'm pretty busy currently.

**Comment by wez** [comment]
In `main`, you can now do something like this:

```lua
local wezterm = require 'wezterm'

return {
  exec_domains = {
    wezterm.exec_domain("woot", function(cmd)
      wezterm.log_info(cmd)
      if cmd.args then
        cmd.args = {
          "bash",
          "-c",
          "echo WOOT && " .. wezterm.shell_join_args(cmd.args)
        }
      end
      return cmd
    end),
  },
  default_domain = "woot",
}
```

You can define multiple `exec_domains` if you wish.

The callback is passed a `SpawnCommand` and the callback must return a `SpawnCommand` that encodes the actual command that will be run.

There are a couple of helper functions that can make some things more convenient:

* [shell_join_args](https://wezfurlong.org/wezterm/config/lua/wezterm/shell_join_args.html)
* [shell_quote_arg](https://wezfurlong.org/wezterm/config/lua/wezterm/shell_quote_arg.html)
* [shell_split](https://wezfurlong.org/wezterm/config/lua/wezterm/shell_split.html)

**Comment by lunaryorn** [comment]
@wez Woa, that's really great.  Can't wait to try it; when will this make it into a release? :innocent:

**Comment by wez** [comment]
I want to get feedback on how well it works before I write up docs and then make a release; can you try a nightly build?

**Comment by lunaryorn** [comment]
Will do next week! ❤️

**Comment by wez** [comment]
Docs should show up at https://wezfurlong.org/wezterm/config/lua/ExecDomain.html shortly

**Comment by lunaryorn** [comment]
@wez I tested nightly, and it works well.  I had to alter the identifier slightly, though: `$WEZTERM_PANE` doesn't seem to change for new tabs, so your code generated the same identifier for every new tab.  However systemd scope identifiers must be unique, so this broke creating new tabs.

I used a random string in place of the `$WEZTERM_UNIX_SOCKET` part, and things worked well.  Can't wait to see this in a release :star_struck:
